### ISSUE TITLE
bug(COMPONENT): CDK Virtual Scroller jump back/flickers to items on top

### ISSUE THREAD

**Issue by Donovantxy** [angular/components#27104]
### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

I'm getting an annoying behaviour while scrolling, where items snap/jump back and I'm never able to scroll all the way down the end of the list. I've notice though, that this bug happens only when I use the virtual scrolling within a material dialog

Template

`
<cdk-virtual-scroll-viewport class="scroll-viewport" itemSize="20">
   <li class="scan-list__item" *cdkVirtualFor="let scan of pointClouds| keyvalue:classesByNameAsc;templateCacheSize:0">
        <div class="scan-list__scan">
            <div class="scan-list__scan__left">
              <div>{{scan.value.name}}</div>
            </div>
        </div>
   </li>
</cdk-virtual-scroll-viewport>
`

CSS

`
 .scroll-viewport {
    height: 200px;
  }
`

### Reproduction

- Implement **cdk-virtual-scroll-viewport** scroller **inside a Angular Material dialog**
- in the template set **templateCacheSize:0**
- try to scroll down

### Expected Behavior

https://github.com/angular/components/assets/6199235/63111f97-c665-4cc0-8f8b-eead8d75a025

### Actual Behavior

While scrolling down, the scroll jump up to the first times

### Environment

- Angular: ~11.2.14
- CDK/Material: ~11.2.13
- Browser(s): Chrome
- Operating System (e.g. Windows, macOS, Ubuntu): macOs Catalina

**Comment by aeslinger0** [comment]
I think when I got this behavior last it was because the itemSize was a tiny bit off of the actual row height. Maybe double check that your itemSize includes border or margins?

**Comment by NikGurev** [comment]
@Donovantxy did you manage to fix the issue?

**Comment by RCout1nho** [comment]
Does anyone have success with this?

**Comment by RCout1nho** [comment]
@Faulkner368 the same happened to me, I think this happens because the cached elements increase the size of the container div and it must interfere with the scroll, I don't know if it's a bug in the lib or if it's possible to solve it by changing the container

**Comment by Faulkner368** [comment]
@RCout1nho was wondering the same, if so I dont think there will be a work around as the component is getting destroyed it cant have any height

**Comment by EmilOlovsson** [comment]
@RCout1nho @Faulkner368 @NikGurev I have spent a few hours with this problem and I found a solution that worked for me.

Browser: Chrome (Version 124.0.6367.118 (Official Build) (64-bit))

Original layout:
`<cdk-virtual-scroll-viewport>
	<div class="cdk-virtual-scroll-content-wrapper"></div>
	<div class="cdk-virtual-scroll-spacer"></div>
</cdk-virtual-scroll-viewport>`

New layout:
`<cdk-virtual-scroll-viewport>
	<div class="cdk-virtual-scroll-spacer"></div>
	<div class="cdk-virtual-scroll-content-wrapper"></div>
</cdk-virtual-scroll-viewport>`

I moved **.cdk-virtual-scroll-spacer** before **.cdk-virtual-scroll-content-wrapper**

**Comment by shiv19** [comment]
I ended up using the `appendOnly` mode

```html
<cdk-virtual-scroll-viewport
            appendOnly
            itemSize="68"
            minBufferPx="1020"
            maxBufferPx="1360">
</cdk-virtual-scroll-viewport>
 ```

**Comment by BdDragos** [comment]
> @RCout1nho @Faulkner368 @NikGurev I have spent a few hours with this problem and I found a solution that worked for me.
> 
> Browser: Chrome (Version 124.0.6367.118 (Official Build) (64-bit))
> 
> Original layout:
> 
> ```
> <cdk-virtual-scroll-viewport>
>     <div class="cdk-virtual-scroll-content-wrapper"></div>
>     <div class="cdk-virtual-scroll-spacer"></div>
> </cdk-virtual-scroll-viewport>
> ```
> 
> New layout:
> 
> ```
> <cdk-virtual-scroll-viewport>
>     <div class="cdk-virtual-scroll-spacer"></div>
>     <div class="cdk-virtual-scroll-content-wrapper"></div>
> </cdk-virtual-scroll-viewport>
> ```
> 
> I moved **.cdk-virtual-scroll-spacer** before **.cdk-virtual-scroll-content-wrapper**

Can you explain how you achieved this please? Since these are auto generated for me when I use cdkVirtualFor, and I also encounter this scrolling issue.

**Comment by Faulkner368** [comment]
@BdDragos I found there were two causes, not having correctly calculated the `itemSize` but this is a well known cause and was not the issue for me and it was caused by using `templateCacheSize:0`.

Setting the cache to zero was not ideal since I want as much of a performance boost as possible so to understand my solution you need to understand why I was setting cache to zero.

I am working with existing, large and complex components that ideally would have been refactored but I did not have time for that. These components were processing and changing property values during construction and / or during initialising life cycle which resulted in an issue where I was seeing the wrong text on screen for a component, this occurs because instead of constructing new components when scrolling it reuses previously constructed components by changing the binded data and during change detection this would get picked up; however the component is an existing one so the constructor and the on initialising method is not executed again so the processing that was needed in my case did not occur and was thus seeing the wrong values.

Setting cache to zero means it would not reuse existing components and would construct new ones, thus solving my issue. However it introduced the issue discussed here and I suspect its due to the calculations on itemSize when components are being destroyed and new ones created there is a brief moment where total height is not adding up... that is my suspicion anyway.

So I did a hacky fix as had insufficient time for a refactor on these large components, so I removed the `templateCacheSize:0` and used OnChanges to do the processing that was taking place originally in OnInit and in the constructor so when the bindings are updated so when change detection runs everything was displayed as it should be.

So in summary I fixed this by removing `templateCacheSize:0` and achieved my goal by using OnChanges to do the processing that was previously happening in OnInit and constructor, hope that helps

**Comment by BdDragos** [comment]
For me the bug happens with and without the templaceCacheSize (but templateCacheSize not to 0 fixes it in most cases*).
Unfortunately in our case ngOnChanges is not a solution since there are some very complex components that need to be rendered and I also depend on other browser limitations that would make the processing not possible without a lot of refactoring and extra code.

**Comment by Faulkner368** [comment]
I hadn't seen this without `templaceCacheSize:0` and most often seemed from stackoverflow and the likes to be caused by element sizes not matching what you have defined in `itemSize`, have you physically checked in the DOM he heights of all the visible elements, I think I initially had issue not having accounted for border

**Comment by BdDragos** [comment]
> I hadn't seen this without `templaceCacheSize:0` and most often seemed from stackoverflow and the likes to be caused by element sizes not matching what you have defined in `itemSize`, have you physically checked in the DOM he heights of all the visible elements, I think I initially had issue not having accounted for border

Yes, I set the itemSize to two decimals accuracy. I also tried to make it "perfect" and intentionally not perfect (small variations), still the same issue.

**Comment by Faulkner368** [comment]
Hard to know then without seeing it first hand, maybe if you have time see if can reproduce in something like https://stackblitz.com/ and share here or Stackoverflow and someone maybe able to help but in the least can help narrow down the cause when in a clean / simple app

**Comment by EmilOlovsson** [comment]
> > @RCout1nho @Faulkner368 @NikGurev I have spent a few hours with this problem and I found a solution that worked for me.
> > Browser: Chrome (Version 124.0.6367.118 (Official Build) (64-bit))
> > Original layout:
> > ```
> > <cdk-virtual-scroll-viewport>
> >     <div class="cdk-virtual-scroll-content-wrapper"></div>
> >     <div class="cdk-virtual-scroll-spacer"></div>
> > </cdk-virtual-scroll-viewport>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > New layout:
> > ```
> > <cdk-virtual-scroll-viewport>
> >     <div class="cdk-virtual-scroll-spacer"></div>
> >     <div class="cdk-virtual-scroll-content-wrapper"></div>
> > </cdk-virtual-scroll-viewport>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I moved **.cdk-virtual-scroll-spacer** before **.cdk-virtual-scroll-content-wrapper**
> 
> Can you explain how you achieved this please? Since these are auto generated for me when I use cdkVirtualFor, and I also encounter this scrolling issue.

@BdDragos 
This should work for you.

// HTML
```

<cdk-virtual-scroll-viewport
	#virtualscroll
	autosize
	[minBufferPx]="A_NUMBER_HERE"
	[maxBufferPx]="A_NUMBER_HERE"
>
	// Your content here
</cdk-virtual-scroll-viewport>
```

// .ts file
```
import {CdkVirtualScrollViewport} from "@angular/cdk/scrolling";

@ViewChild('virtualscroll') private _virtualscrollRef?: CdkVirtualScrollViewport;

ngAfterViewInit() {
	if (this._virtualscrollRef) {
		// @ts-ignore
		const scrollable = this._virtualscrollRef.scrollable.elementRef.nativeElement;
		if (scrollable && scrollable.children) {
			const spacer = scrollable.children[1];
			if (spacer && spacer.classList.contains('cdk-virtual-scroll-spacer')) {
				// Move spacer as first child
				scrollable.insertBefore(spacer, scrollable.firstChild);
			} else {
				console.error("Could not find spacer");
			}
		} else {
			console.error("Could not find scrollable");
		}
	}
}
```

**Comment by BdDragos** [comment]
I succeeded in shifting the elements with a JavaScript type of solution after I thought a bit.

```
    effect(() => {
      const nativElem = this.cdkScrollElement()?.elementRef.nativeElement;
      if (nativElem) {
        const last = nativElem.lastElementChild;
        if (nativElem.firstElementChild && last) {
          nativElem.replaceChild(nativElem.firstElementChild, nativElem.lastElementChild);
          nativElem.insertBefore(last, nativElem.firstElementChild);
        }
      }
    });
```
the above will be put into the constructor(), and it will use a signal based viewChild to trigger as soon as the rendering of the cdkScrollElement is done. The viewChild should be binded directly to the <cdk-virtual-scroll-viewport>

` public cdkScrollElement = viewChild<CdkVirtualScrollViewport | undefined>('cdkScrollElement');
`

This fixed my issue it seems, now I can scroll without problems. Thanks for the suggestion. @shiv19 

> Hard to know then without seeing it first hand, maybe if you have time see if can reproduce in something like https://stackblitz.com/ and share here or Stackoverflow and someone maybe able to help but in the least can help narrow down the cause when in a clean / simple app

It would take too much time to set up everything, the behaviour is pretty complex, but the shifting did it.

**Comment by shiv19** [comment]
@BdDragos 
In my case the actual problem was not having the correct item size, and also rending category headers just inside the for loop even though they were not a part of the list itself.
I changed the category headers to be a part of the list and gave a fixed height for the category rows and the list item rows.
And everything is working fine now.
I even removed the `appendOnly` mode.

This is fine when all the list items can have a fixed height.
When items can't have a fixed height, we'll need to use a custom virtual scroll strategy.

I found a few articles that talk about this:

https://dev.to/georgii/virtual-scrolling-of-content-with-variable-height-with-angular-3a52#final-code

https://angularindepth.com/posts/1091/writing-custom-virtual-scroll-strategy

And here's some code from the CDK experimental:
https://github.com/angular/components/blob/main/src%2Fcdk-experimental%2Fscrolling%2Fauto-size-virtual-scroll.ts

**Comment by canadakickass** [comment]
I can't believe it! switching the spacer fixed the issue for me also. Should this change be brought to the CDK?

**Comment by EdGuan** [comment]
switching the spacer fixed my issue as well. Im currently using Angular 17, and my issue is not caused by the `itemSize` and `templaceCacheSize:0`.

**Comment by radek-vizlib** [comment]
I'm working with the `{ ngZone: 'noop' }` setting, so I need to run detectChanges manually.
I have the same issue after upgrading Angular from 13 to 18.
The `appendOnly` workaround works, but I do not want to use it.

For virtual scrolling, I've added a manual call to `_doChangeDetection`:
```js
    public onScrolledIndexChange(): void {
        this.viewPort['_doChangeDetection']();  // manual _doChangeDetection fix
        this.changeDetection.detectChanges();
    }
```
and it works as expected.

**Comment by mhaertwig** [comment]
I spent some time debugging the jumping `CdkVirtualScrollViewport` and want to share my findings here.

I am using `angular/cdk` version `19.1.4`.

This [Stackblitz](https://stackblitz.com/edit/e7wqrpn1?file=src%2Fexample%2Fcdk-virtual-scroll-overview-example.ts,src%2Fexample%2Fcdk-virtual-scroll-overview-example.html,src%2Fexample%2Fcdk-virtual-scroll-overview-example.css) shows the jumping scroll bar.

Here's how to fix it.

1. **Fix imports**: Instead of importing the `CdkVirtualScrollViewport`, `CdkVirtualForOf` and `CdkFixedSizeVirtualScroll`, just import the `ScrollingModule`. The components seem to interfere with each other causing the scroll bar to jump.
2. **Remove the "scroll to bottom" button**: The button is constantly repainted because its `disabled` state is based on a function call. This seems to make the scroll viewport forget it's last position.

**Comment by Frtrillo** [comment]
> I spent some time debugging the jumping `CdkVirtualScrollViewport` and want to share my findings here.
> 
> I am using `angular/cdk` version `19.1.4`.
> 
> This [Stackblitz](https://stackblitz.com/edit/e7wqrpn1?file=src%2Fexample%2Fcdk-virtual-scroll-overview-example.ts,src%2Fexample%2Fcdk-virtual-scroll-overview-example.html,src%2Fexample%2Fcdk-virtual-scroll-overview-example.css) shows the jumping scroll bar.
> 
> In my case, I was able to fix the jumping scrollbar by doing the following:
> 
> 1. **Fix imports**: Instead of importing the `CdkVirtualScrollViewport`, `CdkVirtualForOf` and `CdkFixedSizeVirtualScroll`, just import the `ScrollingModule`. The components seem to interfere with each other causing the scroll bar to jump.
> 2. **Remove the "scroll to bottom" button**: The button is constantly repainted because its `disabled` state is based on a function call. This seems to make the scroll viewport forget it's last position.

im using only ScrollingModule and still got this issue

**Comment by MorenoAlex** [comment]
> I succeeded in shifting the elements with a JavaScript type of solution after I thought a bit.
> 
> ```
>     effect(() => {
>       const nativElem = this.cdkScrollElement()?.elementRef.nativeElement;
>       if (nativElem) {
>         const last = nativElem.lastElementChild;
>         if (nativElem.firstElementChild && last) {
>           nativElem.replaceChild(nativElem.firstElementChild, nativElem.lastElementChild);
>           nativElem.insertBefore(last, nativElem.firstElementChild);
>         }
>       }
>     });
> ```
> 
> the above will be put into the constructor(), and it will use a signal based viewChild to trigger as soon as the rendering of the cdkScrollElement is done. The viewChild should be binded directly to the <cdk-virtual-scroll-viewport tag
> 
> `public cdkScrollElement = viewChild<CdkVirtualScrollViewport | undefined>('cdkScrollElement');`
> 
> This fixed my issue it seems, now I can scroll without problems. Thanks for the suggestion. [@shiv19](https://github.com/shiv19)
> 
> > Hard to know then without seeing it first hand, maybe if you have time see if can reproduce in something like https://stackblitz.com/ and share here or Stackoverflow and someone maybe able to help but in the least can help narrow down the cause when in a clean / simple app
> 
> It would take too much time to set up everything, the behaviour is pretty complex, but the shifting did it.

Unfortunatly this doesn't seem to work with Angular 19 anymore.

**Comment by dzolotarova** [comment]
> [@RCout1nho](https://github.com/RCout1nho) [@Faulkner368](https://github.com/Faulkner368) [@NikGurev](https://github.com/NikGurev) I have spent a few hours with this problem and I found a solution that worked for me.
> 
> Browser: Chrome (Version 124.0.6367.118 (Official Build) (64-bit))
> 
> Original layout:
> 
> ```
> <cdk-virtual-scroll-viewport>
>     <div class="cdk-virtual-scroll-content-wrapper"></div>
>     <div class="cdk-virtual-scroll-spacer"></div>
> </cdk-virtual-scroll-viewport>
> ```
> 
> New layout:
> 
> ```
> <cdk-virtual-scroll-viewport>
>     <div class="cdk-virtual-scroll-spacer"></div>
>     <div class="cdk-virtual-scroll-content-wrapper"></div>
> </cdk-virtual-scroll-viewport>
> ```
> 
> I moved **.cdk-virtual-scroll-spacer** before **.cdk-virtual-scroll-content-wrapper**

I also found several bugs in my projects with virtual scroll (unfortunately, it's not possible to reproduce these on stackblitz). The scroll bar jumps up slightly when:
1. you scroll almost to the end of the list
2. you add new items for the virtual scroll

Replacing the wrapper with a spacer helped me.

**Comment by lppedd** [comment]
Folks, my case is - yet again - slightly different. I've found out that using integer values (no decimals) for the item height solves the problem. Obviously this means that the scrolling isn't pixel-precise, but it seems to solve the jumping issue.

**Comment by lppedd** [comment]
Note that while using `appendOnly` may fix the issue, it goes against the idea of virtual scrolling: eliminating unnecessary DOM nodes.

**Comment by GowthamKrishna362** [comment]
Hey, not sure if this is what you are looking for, but i had a similar issue, and fixed it by giving a fixed height to the child item of the scroll list
```
<cdk-virtual-scroll-viewport
  [itemSize]="itemSize"
>
<div cdkVirtualFor="..." [style.max-height.px]="itemSize">
 .... Your content here
</div>
```

This fixed it. Might have to add width instead for horizontal scroll, and there might be other similar conditions, but this seems to have solved my issues

**Comment by ssc-io** [comment]
Hi together, only moving the spacer before the wrapper was not enough for me.

If i wanted to scroll to the very last row, **the viewport scrolled only to the pre-last row**.
(Or to be more precise, the scrollbar jumped back to the pre-last row).

But i found a solution 🎉. The problem was caused by the sticky header.
The spacer is just not high enough to scroll to the very bottom. So i needed to adjust it.

Here is what i did:
```ts
host: {
    '[style.--header-row-height]': 'headerRowHeightPx()', // define a signal for dynamic header row height
},
```
```scss
cdk-virtual-scroll-viewport ::ng-deep.cdk-virtual-scroll-spacer {
  margin-top: calc(var(--header-row-height) * 1px);
}
```

And that's it. It can now scroll to the very bottom.
I hope i can help anyone ❤

**Comment by ADH-LukeBollam** [comment]
Maybe it's been fixed in newer versions but as the issue is still open I'll add my fix here. It seems to be a conflict with the browser scroller behavior trying to snap the scroller at least in v17. Add this to your viewport css to prevent the browser anchoring and snapping:

```css
.viewport {
  /* Important!! these prevent a bug where the scroller snaps / jumps backwards during scrolling */
  overflow-anchor: none;
  scroll-behavior: auto;
  scroll-snap-type: none;
}

.viewport .cdk-virtual-scroll-content-wrapper {
  overflow-anchor: none;
}
```
